### ISSUE TITLE
Misc link picker modal issues

### ISSUE THREAD

**Issue by abjerner** [umbraco/Umbraco-CMS#16838]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

14.1

### Bug summary

I've tried playing a bit around with the link picker modal, and have experienced a few issues, which resulted in me starting a [discussion on Discord](https://discord.com/channels/869656431308189746/882984798719729704/threads/1266006511046627390).

Some of my issues may be due to lack of documentation, as I haven't been able to find any documentation or examples on how to use the modal.

#### No import map
Most modals seem to be a part of `@umbraco-cms/backoffice/modal`, but the link picker modal is not part of this import, nor any other imports. Lee Kelleher mentioned that this will addressed for 14.2, so looking forward to this.

#### Option to hide anchor doesn't work
It seems that there is an option to hide [anchor field](https://github.com/umbraco/Umbraco.CMS.Backoffice/blob/v14.1.0/src/packages/multi-url-picker/link-picker-modal/link-picker-modal.token.ts#L14), but I can't work out how to use this. The comment above the referenced line also suggests that this may not work.

#### No option to hide target property
The link picker modal/overlay has a **Target** property. In previous versions of Umbraco, it was possible to hide this option for context where a target doesn't really make sense. I'm using this in **Skybrud Redirects**, so I'm missing this option in Umbraco 14.

#### No option to restrict link type
In Umbraco 6 and uComponents, the URL picker had an option to restrict the type of link that could be submitted. E.g. to so that only content and media links would be allowed, but not external links. I'd imagine the Multi URL picker was somewhat inspired by the URL picker in uComponents, and I've been missing this option a few times for the Multi URL picker.

It would be great for both the Multi URL picker property editor and the link picker modal to support this.

#### Editing link title or query/anchor removes `unique` and `url` from submitted value
In Umbraco 13 and earlier versions, one could select a content or media item, in which case the **Link title** property would be populated with the name of the selected item. One could then afterwards edit both the link title and anchor/query, which would then be part of the submitted link object.

In Umbraco 14, editing either of the **Anchor / querystring** or **Link title** properties, the item selection is lost, meaning the `unqiue` and `url` properties are missing in the submitted link object.

#### `url` of submitted value is the GUID key, not the actual URL
In previous versions of Umbraco, selecting a content or media item, the **Link** field would be populated with the URL of that item, and also be the value of the `url` property of the resulting link object.

In Umbraco 14 the field isn't populated with the URL, and the `url` property is simply the GUID key of the selected item, not the URL.

#### Setting initial link value doesn't work
In Umbraco 14, setting an initial link doesn't seem to be supported. Since submitting the link picker modal returns in a value object with a `link` property, I would imagine something like this would set the initial link that is shown when the picker is opened:

```js
const modalContext = this.modalManagerContext?.open(this, UMB_LINK_PICKER_MODAL, {
    modal: {
        size: "medium",
    },
    value: {
        link: {
            type: "document",
            unique: "b87dc2bb-26ae-48f3-93fc-1b9b58b38c62",
            url: "b87dc2bb-26ae-48f3-93fc-1b9b58b38c62"
        }
    }
});
```

### General layout
The general layout of the overlay/modal could IMO be improved. For one, the default size of the modal is `small`, result in the modal components being very compact. For custom code opening the modal, one can always set the size to something like `medium` instead, but I'd say that the **Link** and **Anchor / querystring** ideally should be laid out on two different lines, rather than on the same line.

![image](https://github.com/user-attachments/assets/d347e89f-dbe1-489c-84d6-1a0d24c552f5)

![image](https://github.com/user-attachments/assets/91772ef1-3457-4356-864b-74674d09a59c)

### Steps to reproduce

See *Bug summary*.

---
_This item has been added to our backlog AB#44594_

**Comment by bjarnef** [comment]
Regarding Anchor / querystring I think it is fine these fields are next to each other, but it can use CSS Container Queries https://caniuse.com/css-container-queries so the two fields are stacked in small modal width.

**Comment by leekelleher** [comment]
@abjerner Apologies for the delayed feedback on these issues you raised a while ago.

I'll take each one in turn...

**No import map**

As of v14.2.0, the link picker modal is within the "multi-url-picker" package/namespace, (this was moved from the general "core/modal" to its own module/package for bundling purposes).

It can be imported with...

```
import { UMB_LINK_PICKER_MODAL } from "@umbraco-cms/backoffice/multi-url-picker";
```

---

**Option to hide anchor doesn't work**

The `hideAnchor` flag can be passed into the modal via `data.config`.

```
this.modalManagerContext?.open(this, UMB_LINK_PICKER_MODAL, { data: { config: { hideAnchor: true } } });
```

Of course, documentation around these types of configurations is currently lacking. We can fix in due course.

---

**No option to hide target property**

This option is missing from v14/v15, I'll look to add it back, in a similar way to `hideAnchor`, e.g. `hideTarget`.
Currently estimated for Umbraco v15.1, (as v15.0-rc is feature locked).

---

**No option to restrict link type**

This is definitely open for consideration and further discussions.


**Editing link title or query/anchor removes `unique` and `url` from submitted value**

I haven't been able to reproduce this specific issue in the latest codebase (v15-rc2 at the time of writing), but there is definitely some further work to be done with the UI when selecting a document or media item and having the link title populated (if empty).

---

**`url` of submitted value is the GUID key, not the actual URL**

This is a known bug, I'm tasked with fixing this.

---

**Setting initial link value doesn't work**

Your code snippet looks correct to me, this should work.

At least that's how we're doing it in the core, [see the TinyMCE link picker plugin code](https://github.com/umbraco/Umbraco.CMS.Backoffice/blob/v15.0.0-rc2/src/packages/multi-url-picker/tiny-mce-plugin/tiny-mce-multi-url-picker.plugin.ts#L80-L88).

---

**General layout**

I agree that the link picker is better suited to the "medium" modal size. I'll review the URL/Anchor/Querystring inputs.

---

My aim is to review all of these issues during the development cycle for v15.1.0, so hopefully will enable you to make progress with Skybrud Redirects.

**Comment by leekelleher** [comment]
The majority of these issues have been addressed by https://github.com/umbraco/Umbraco.CMS.Backoffice/pull/2477, which will be available in Umbraco v15.1.0.

Anything specific that has been missed (e.g. restricting to only document/media), can be opened as a new feature request.
